lib/materialist/materialized_record.rb

```diff
@@ -8,12 +9,20 @@ def self.included(base)
     end
 
     module ClassMethods
-      def source_link_reader(*keys, via: nil)
+      def source_link_reader(*keys, via: nil, allow_nil: false)
         keys.each do |key|
           define_method(key) do
-            raw = source_raw
-            raw = raw.send(via).show if via
-            raw.send(key).show.body
+            (via ? [via, key] : [key])
+              .inject(source_raw) do |res, path|
+                begin
+                  (res && res.body._links.include?(path)) ?
+                    res.send(path).show :
+                    (allow_nil ? nil : raise(ResourceNotFound))
+                rescue Routemaster::Errors::ResourceNotFound
+                  (allow_nil ? nil : raise(ResourceNotFound))
+                end
+              end
```

Comment:
I find this section quite hard to parse - might be clearer with explicit `if` statements / extracted to its own method.

Reply:
@robladbrook @knaveofdiamonds part of the reason this is left in it's convoluted form is to avoid adding any helper method here because any helpers here would be accessible through the materializer types.

It's not one of those things that you would come to read everyday, so I would leave it alone 😄